### ISSUE TITLE
Config file triage and consolidation

### ISSUE THREAD

**Issue by r2d4** [kubernetes/minikube#1814]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): feature request

Currently we maintain three separate config files.  

* `$MINIKUBE_HOME/config/config.json` **The "viper" config**
This is a global config file that is modified by the `minikube config` command.  Takes precedence over default values.  Is used across all commands.  It also controls addon-management, which today cannot be toggled on a per-profile basis.      

```
$ cat ~/.minikube/config/config.json 
{
    "WantReportErrorPrompt": false,
    "WantUpdateNotification": false,
    "vm-driver": "kvm"
}
```

* `$MINIKUBE_HOME/profiles/$PROFILE_NAME/config.json` **the "kubernetes config"**
A combination of the machine configuration and the kubernetes options.  Saved or overwritten on `minikube start`.  Currently used to store the profile's current kubernetes configuration.

```
cat ~/.minikube/profiles/minikube/config.json
{
    "MachineConfig": {
        "MinikubeISO": "https://storage.googleapis.com/minikube/iso/minikube-v0.23.0.iso",
        "Memory": 2048,
        "CPUs": 4,
        "DiskSize": 20000,
        "VMDriver": "kvm",
        "XhyveDiskDriver": "ahci-hd",
        "DockerEnv": null,
        "InsecureRegistry": [
            "10.0.0.0/24"
        ],
        "RegistryMirror": null,
        "HostOnlyCIDR": "192.168.99.1/24",
        "HypervVirtualSwitch": "",
        "KvmNetwork": "default",
        "DockerOpt": null,
        "DisableDriverMounts": false
    },
    "KubernetesConfig": {
        "KubernetesVersion": "v1.7.2",
        "NodeIP": "192.168.39.11",
        "APIServerName": "minikubeCA",
        "DNSDomain": "cluster.local",
        "ContainerRuntime": "",
        "NetworkPlugin": "",
        "FeatureGates": "",
        "ExtraOptions": null
    }
}
```

* `$MINIKUBE_HOME/machines/$MACHINE_NAME/config.json` **the "docker-machine" config**
The docker-machine configuration.  Different than our `MachineConfig`, it is the libmachine config which holds the driver and host structs.  This config is used to store hypervisor driver information, so we can load the correct driver for running machines.  

```
$ cat ~/.minikube/machines/minikube/config.json 
{
    "ConfigVersion": 3,
    "Driver": {
        "IPAddress": "192.168.39.13",
        "MachineName": "minikube",
        "SSHUser": "",
        "SSHPort": 22,
        "SSHKeyPath": "",
        "StorePath": "/usr/local/google/home/mrick/.minikube",
        "SwarmMaster": false,
        "SwarmHost": "",
        "SwarmDiscovery": "",
        "Boot2DockerURL": "file:///usr/local/google/home/mrick/.minikube/cache/iso/minikube-v0.23.1.iso",
        "PrivateKeyPath": "",
        "CPU": 2,
        "Memory": 2048,
        "DiskSize": 20000,
        "NetworkName": "minikube-net",
        "DiskPath": "/usr/local/google/home/mrick/.minikube/machines/minikube/minikube.img",
        "ISO": "/usr/local/google/home/mrick/.minikube/machines/minikube/boot2docker.iso",
        "CacheMode": "default"
    },
    "DriverName": "kvm",
    "HostOptions": {
        "Driver": "",
        "Memory": 0,
        "Disk": 0,
        "EngineOptions": {
            "ArbitraryFlags": null,
            "Dns": null,
            "GraphDir": "",
            "Env": null,
            "Ipv6": false,
            "InsecureRegistry": [
                "10.0.0.0/24"
            ],
            "Labels": null,
            "LogLevel": "",
            "StorageDriver": "",
            "SelinuxEnabled": false,
            "TlsVerify": false,
            "RegistryMirror": null,
            "InstallURL": ""
        },
        "SwarmOptions": {
            "IsSwarm": false,
            "Address": "",
            "Discovery": "",
            "Agent": false,
            "Master": false,
            "Host": "tcp://0.0.0.0:3376",
            "Image": "swarm:latest",
            "Strategy": "spread",
            "Heartbeat": 0,
            "Overcommit": 0,
            "ArbitraryFlags": null,
            "ArbitraryJoinFlags": null,
            "Env": null,
            "IsExperimental": false
        },
        "AuthOptions": {
            "CertDir": "/usr/local/google/home/mrick/.minikube",
            "CaCertPath": "/usr/local/google/home/mrick/.minikube/certs/ca.pem",
            "CaPrivateKeyPath": "/usr/local/google/home/mrick/.minikube/certs/ca-key.pem",
            "CaCertRemotePath": "",
            "ServerCertPath": "/usr/local/google/home/mrick/.minikube/machines/server.pem",
            "ServerKeyPath": "/usr/local/google/home/mrick/.minikube/machines/server-key.pem",
            "ClientKeyPath": "/usr/local/google/home/mrick/.minikube/certs/key.pem",
            "ServerCertRemotePath": "",
            "ServerKeyRemotePath": "",
            "ClientCertPath": "/usr/local/google/home/mrick/.minikube/certs/cert.pem",
            "ServerCertSANs": null,
            "StorePath": "/usr/local/google/home/mrick/.minikube"
        }
    },
    "Name": "minikube"
}
```

We should have one source of truth for minikube configuration management.  We should merge all these configs in a 2 phase merge.

- [ ] **Phase 1:**  Merge viper config and kubernetes config

The viper config currently isn't profile aware.  However, I like the idea of being able to specify global configuration.  I think that the `minikube config` command should respect the `--profile` flag and write the viper config to that profile.  The machine + kubernetes config together is a subset of the viper config.  The viper config contains additional data (such as WantUpdateNotification and log level) as well as addon information.  A few things that might be necessary for merging

* `pkg/minikube/config` must be moved to `cmd/minikube/config`.  No configuration code should be in `pkg/**`.  Right now its needed because the addon configuration isn't passed through the KubernetesConfig.  Any information needed must be passed through MachineConfig or KubernetesConfig.
* A MachineConfig and KubernetesConfig should be able to be generated from a global, then profile config.json.  Maybe we should add a another struct that stores all the neither Machine nor Kubernetes config options (like WantUpdateNotifications and log level).
* `minikube config` should respect `--profile`.  That is, `minikube config set vm-driver kvm --profile cluster2` should write to a profile specific config file.
* Configuration in order of descending precedence should be:
 Flag
 Environment Variable
 Profile Config
 Global Config
 Default Value

At the end of this phase, we'll have two config files of the same schema which will be able to generate a MachineConfig and a KubernetesConfig.  The docker-machine config will also still exist

`~/.minikube/profiles/minikube/config.json` profile configuration
`~/.minikube/config/config.json` global configuration

- [ ] **Phase 2** : Merge combined viper/k8s config with docker-machine config.

This will require us to write our own implementation of the libmachine filestore.  https://github.com/kubernetes/minikube/blob/master/pkg/minikube/machine/client.go#L66

I don't think this would be too difficult, since we already use a custom libmachine API implementation.  The Store interface is very simple and we already have the code lying around:
https://github.com/docker/machine/blob/master/libmachine/persist/store.go#L7-L22

At the end of this phase, we'll have 

cc @darkowlzz as they expressed some interest in implementing something like this.

I think the next step would be to agree on a more formal design before we start implementing anything.  This was more of a stream of consciousness about our current configuration system.

**Comment by darkowlzz** [comment]
I spent some time thinking and prototyping (to aid the thoughts) the Phase 1 of this.

To combine viper config and profile config into one:
* I started with generation of `MachineConfig` and `KubernetesConfig` from `MinikubeConfig`. I wrote two loaders, `loadMachineConfig` and `loadKubernetesConfig`. Given a `MinikubeConfig`, the loaders would loads only the relevant properties to `MachineConfig` and `KubernetesConfig`.
* From above, I realized that instead of writing a huge custom MachineConfig + KubernetesConfig structure as profile config, just writing the essential properties with equivalent MinikubeConfig property name to profile config is all we need. Simplifying the config from the huge JSON that exists now.
So now, the profile config and global config have the same schema ✅ 

After simplifying the profile config, new profile config would look like this:
```js
{
    "kubernetes-version": "v1.7.4"
}
```
Just the k8s version, which is required to avoid version downgrades.

The above work is under a WIP PR #1871.

I'll post more updates as they happen. Feedback would be great 🙂

**Comment by darkowlzz** [comment]
Updates:

* In order to have config hierarchy with overriding at global, profile, environment vars and flags levels, I created a `MinikubeConfig` variable with all the default values (Machine config, k8s config, notice configs, add-ons config). This default config variable can be obtained from `getDefaultConfig()`, see #1871 .
* Once this default config is loaded, the global config is read and the properties defined in global config override the default config values.
* Next, the profile config of the current profile is read and these properties override the previous config.
* Similarly, environment var configs and flags based properties can override to give the final minikube config.
* This combined minikube config is then used to generate a `MachineConfig` and `KubernetesConfig`, which are used to load VM and start k8s cluster.

Some concerns:
* The viper configs are flags with some default values. Flags being at the top of all other configs, overrides all the other configs. Even the unused flags end up overriding configs with their default values. In this way, the properties in config files would never be used 🙁 I need some help with this. Design blocker? 

@r2d4 sorry for starting some implementation without agreeing on a formal design first 😅 I was finding it really hard to think without some code. This prototyping helped in thinking and understanding the issues better 🙂

**Comment by misakwa** [comment]
@r2d4 this seems very interesting and might be the underlying cleanup required to allow creating a development environment very easily from a config file for a team. 
Using the same setup for team environments currently requires a script with all the start options and some commands after the vm is up to configure addons and such.

I had initially thought about wrapping the start command to take a config file, but  haven't dug deeper into the implementation to know how many places need to be updated to ensure the configs pass through to the right components.

I am definitely interested in this cleanup and its benefits to making it easier to write utilities around minikube.

#### TL;DR
- I am interested in using a config file to boostrap a minikube environment for a team
- The point above might require some cleanup in the area of config management like is being suggested by @r2d4
